<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [Mono-list] exceptions and ECMA issues
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:lupus%40debian.org">
   <META NAME="robots" CONTENT="index,nofollow">
   
   <LINK REL="Previous"  HREF="001328.html">
   <LINK REL="Next"  HREF="001368.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[Mono-list] exceptions and ECMA issues
   </H1>
    <B>Paolo Molaro
    </B> 
    <A HREF="mailto:lupus%40debian.org"
       TITLE="[Mono-list] exceptions and ECMA issues">lupus@debian.org
       </A><BR>
    <I>Wed, 22 Aug 2001 17:35:40 +0200</I>
    <P><UL>
        <LI> Previous message: <A HREF="001328.html">[Mono-list] exceptions and ECMA issues
</A></li>
        <LI> Next message: <A HREF="001368.html">[Mono-list] exceptions and ECMA issues
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#1337">[ date ]</a>
              <a href="thread.html#1337">[ thread ]</a>
              <a href="subject.html#1337">[ subject ]</a>
              <a href="author.html#1337">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>On 08/22/01 Nick Drochak wrote:
&gt;<i> &gt; 2) rethrow opcode. It is not clear where the search for an handler
</I>&gt;<i> &gt; should begin after a rethrow: should it start from the beginning,
</I>&gt;<i> &gt; from the last handler found or from the parent method?
</I>&gt;<i> &gt; Consider:
</I>&gt;<i> &gt;
</I>&gt;<i> -- snip code example --
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; Restarting from the beginning will result in an infinite loop,
</I>&gt;<i> &gt; so, what other option should I choose?
</I>&gt;<i> 
</I>&gt;<i> &gt;From the C# Language Spec (sec 8.9.5 The throw statement):
</I>&gt;<i> &quot;When an exception is thrown, control is transferred to the first catch
</I>&gt;<i> clause in an enclosing try statement that can handle the exception. &quot;
</I>
Ok, what I missed is that there is a new throw point (at the rethrow
opcode), so I start the search for an handler from that point.
Rethrow should work in the next snapshot.

&gt;<i> The operative word being &quot;enclosing.&quot; Control shouldn't transfer to any
</I>&gt;<i> catch block in the same try, it should always propagate outwards.
</I>&gt;<i> 
</I>&gt;<i> There is a excruciatingly detailed process description of exeception
</I>&gt;<i> propagation in section 8.9.5 of the spec.  Perhaps that would help?
</I>
Actually, section 15.9.10 confuses another issue a bit, I think.
It says:

--snip snip--
*  If an exception is propagated to the try statement during execution of the try block: 
	o The catch clauses, if any, are examined in order of appearance to locate a suitable handler for the 
exception. The first catch clause that specifies the exception type or a base type of the exception type 
is considered a match. A general catch clause is considered a match for any exception type. If a 
matching catch clause is located: 
		*  If the matching catch clause declares an exception variable, the exception object is assigned to the 
exception variable. 
		*  Control is transferred to the matching catch block. [1]
		*  When and if control reaches the end point of the catch block: 
			o If the try statement has a finally block, the finally block is executed. [2]
			o Control is transferred to the end point of the try statement. 
--snip snip--

Basically it says that if an exception id thrown, control is transfered to the
matching catch block [1] and only later to the finally block [2].
Partition I, instead, claims that a matching catch block is searched, but
before executing it, control is transfered to the finally block
(section 11.4.2.2.):

--snip snip--
If a match is not found in the current method, the calling method is searched, and so on. If no match is found 
the CLI will dump a stack trace and abort the program. If a match is found, the CLI walks the stack back to the 
point just located, but this time calling the finally and fault handlers. It then starts the corresponding exception 
handler. Stack frames are discarded either as this second walk occurs or after the handler completes, depending 
on information in the exception handler array entry associated with the handling block. 
--snip snip--

I implemented what is specced out in Partition I, hope the ECMA folks can clarify
this issue.

lupus

-- 
-----------------------------------------------------------------
<A HREF="mailto:lupus@debian.org">lupus@debian.org</A>                                     debian/rules
<A HREF="mailto:lupus@ximian.com">lupus@ximian.com</A>                             Monkeys do it better


</PRE>
<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI> Previous message: <A HREF="001328.html">[Mono-list] exceptions and ECMA issues
</A></li>
	<LI> Next message: <A HREF="001368.html">[Mono-list] exceptions and ECMA issues
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#1337">[ date ]</a>
              <a href="thread.html#1337">[ thread ]</a>
              <a href="subject.html#1337">[ subject ]</a>
              <a href="author.html#1337">[ author ]</a>
         </LI>
       </UL>
</body></html>
